### ISSUE TITLE
[Feature] Fog Affecting Skies

### ISSUE THREAD

**Issue by MajorCooke** [ZDoom/gzdoom#2696]
Seeing [this PR](https://github.com/ZDoom/gzdoom/pull/2695) has me thinking of fog affecting skies, such as bright fog often occludes the skies, while a clear dark night with moonlight still has silhouettes.

This could also benefit from being toggleable in both cvar and MAPINFO, with the cvar allowing overriding for the sake of testing.

**Comment by dileepvr** [comment]
The fact the `skyfog` exists as a parameter (the amount of fog density to apply to skies, with the color taken from the sector's fog) tells me that attempts at making an automatic solution work for all skies were given up on, and control was handed over to the map maker. 
I am also reading that `skyfog` does not affect six-sided skyboxes. Only wrap-around single-texture skies.

**Comment by MajorCooke** [comment]
I think `skyfog` is something to do with looking straight up and down? I'm not sure. 

Do you think this could be possible akin to the Unreal way or no? Basically akin to it being a cylinder. Without it, the pop-ins for BSP, while understandable, may be really jarring. And again, I'm not worried about it until after your PR is merged.

**Comment by dileepvr** [comment]
Since adding new vertices, etc. to the levelmesh after the BSP nodes have been generated is a no no, the cylinder will likely have to be a parallel levelmesh (is that the right term?) or a giant, inside-out 3D model that surrounds and moves with the  viewpoint. The latter might be testable right now. I might do it and report back here.

**Comment by MajorCooke** [comment]
Fantastic, looking forward to it.

**Comment by dileepvr** [comment]
Having looked at how `skyviewpoint` actor functions in, say, here: https://www.realm667.com/repository/sfx-shoppe/other/2397-parallaxing-sky-viewpoint#preview

It seems to me that a gradient texture may be erected on linedefs just as the tree-line or mountains were in the skyviewpoint area.

**Comment by MajorCooke** [comment]
I know. But it'd still be nice to be able to do this through *other* means for dynamic fog.

**Comment by MajorCooke** [comment]
Also that doesn't solve the problem of having it not being able to affect skies without skyboxes.

**Comment by inkoalawetrust** [comment]
Yeah I agree, this should definitely be a feature, having fog blend into the sky. Not sure what my (Well actually kinda DrPyspy's, he wrote the actual code for moving around the camera by some division factor back in 2019 when I could code anything, but I was the one that expanded it to be more useful afterwards) moving sky camera is supposed to prove.
What do you mean by gradient texture? Like a fog gradient in the background of a skybox? That's still not really a solution. Even in skybox sectors.

**Comment by MajorCooke** [comment]
WITH the caveat that it can be toggled as well. Since, you know, dark skies can be seen through the dark.

**Comment by MajorCooke** [comment]
>  What do you mean by gradient texture? Like a fog gradient in the background of a skybox?

Look at the second link in the OP.

**Comment by inkoalawetrust** [comment]
Yeah, that's what I thought too. But as I said I think it's a band aid solution, on top of of course being only really usable in skybox sectors.

**Comment by dileepvr** [comment]
Bump. So the first 12 indices in the vbo cover 12 triangles that form a box around the viewpoint, right? The code to apply the skyfog right now just runs the following lines when drawing the first sky portal:

`PalEntry FadeColor = origin->fadecolor;`
`FadeColor.a = clamp<int>(di->Level->skyfog, 0, 255);`

`state.EnableTexture(false);`
`state.SetObjectColor(FadeColor);`
`state.Draw(DT_Triangles, 0, 12);`
`state.EnableTexture(true);`
`state.SetObjectColor(0xffffffff);`

(I was able to add this to six-sided skyboxes too btw. Don't know why it was kept out.)

To paint a cylinder with a pitch-gradient skyfog we'd just need to store a dense mesh of triangles representing the cylinder at a known fixed range of vboindices, arranged such that banded rings of triangles on the cylinder can be painted with one intensity before going up to the next ring. Right?

Where are the first 12 triangles defined?

**Comment by dileepvr** [comment]
Never find. Found it. It was in `hw_skydome.cpp` line 289 onwards.
This feature is trivial if someone doesn't mind hard-coding some dense triangle meshes for a unit-sphere in ordered, banded rings for latitudes.

**Comment by dileepvr** [comment]
Gah. There already is a hemisphere vertices create function ... for the vanilla doom sky! Only for rows (`mRows`) though. ...

**Comment by MajorCooke** [comment]
> Nevermind. Found it. It was in `hw_skydome.cpp` line 289 onwards. This feature is trivial if someone doesn't mind hard-coding some dense triangle meshes for a unit-sphere in ordered, banded rings for latitudes.

Best of luck to any who do, because I have no idea how to handle this, nor do I have time to do so.

**Comment by dileepvr** [comment]
Got it working: https://www.youtube.com/watch?v=H_zX09ye0Pc

I am packing a `SKYMIST1` texture lump (png file) of size 1x544 in `gzdoom.pk3` that is all white, with a linear alpha ramp from the midpoint to the top. Use mod-able, of course. Is this the feature being requested?

**Comment by MajorCooke** [comment]
**Yes!**

**Comment by MajorCooke** [comment]
Yes! But I'm also wondering if this could also be made to work going down as well? And perhaps with the option to work both ways?

**Comment by MajorCooke** [comment]
![Image](https://github.com/user-attachments/assets/223bc0c9-ce87-4734-9a62-6abce496ecaa)
It would allow for things like this to be made more consistently too.

**Comment by dileepvr** [comment]
You can provide that texture as a `SKYMIST1` lump replacement and it would work. But I suppose I can throw in a `SKYMIST2` and a `SKYMIST3` into `gzdoom.pk3` to give modders three ready-made options. They will all be 1x544 pixels, 8-bit grayscale pngs with alpha channel.

**Comment by MajorCooke** [comment]
Works for me, but if anything I might go as far as suggest letting them define via mapinfo what the graphic lump should be for each map, or specifying a global default.

**Comment by coelckers** [comment]
This hadn't been closed along with the PR.
